### ISSUE TITLE
mycms and YubiHSM2

### ISSUE THREAD

**Issue by gilweis** [alonbl/mycms#10]
YubiHSM2 requires login to see the objects in the token. Without performing a login the object list is empty including public and certificate type objects.
When trying to perform a signature, the operation fails because the certificate is not found.

> ERROR: 00100004: No object found
> MYCMS ERROR DUMP - BEGIN
> Version: mycms-0.2.0_master ()
> Entries (most recent call last):
>     #0
>         SOURCE_FILE=mycms-certificate-driver-pkcs11.c
>         SOURCE_LINE=356
>         SOURCE_FUNC=__find_object
>         HINT=certificate.driver.pkcs11.find
>         CODE=0x00100004
>         AUTHORITATIVE=1
>         DESCRIPTION=No object found
> MYCMS ERROR DUMP - END

**Comment by alonbl** [comment]
On Sat, Feb 3, 2024 at 7:16 PM gilweis ***@***.***> wrote:

> YubiHSM2 requires login to see the objects in the token. Without
> performing a login the object list is empty including public and
> certificate type objects.
> When trying to perform a signature, the operation fails because the
> certificate is not found.
>

You should be able to set certificate as public when creating the object
just like the other providers.Message ID: ***@***.***
com>
>
>

**Comment by gilweis** [comment]
With YubiHSM2  it's Impossible to access pubic keys and certificates objects without login.
No problem with softhsm or yubikey to access them without login.

```
$ pkcs11-tool --module /usr/lib/x86_64-linux-gnu/pkcs11/yubihsm_pkcs11.so -O  --pin 0001password
Private Key Object; RSA 
  label:      label_perm2
  ID:         0066
  Usage:      decrypt
  Access:     sensitive, always sensitive, extractable, local
  Allowed mechanisms: RSA-PKCS,RSA-PKCS-OAEP
Public Key Object; RSA 2048 bits
  label:      label_perm2
  ID:         0066
  Usage:      encrypt
  Access:     local
```

```
$ pkcs11-tool --module /usr/lib/x86_64-linux-gnu/pkcs11/yubihsm_pkcs11.so -O
Using slot 0 with a present token (0x0)

```

**Comment by alonbl** [comment]
I see that they support[1] CKA_PRIVATE attribute, so if you store the
certificate object you can control the attribute. A provider that does not
allow public access to certificate will not work with most of the
applications, so it is strange.

[1]
https://docs.yubico.com/hardware/yubihsm-2/hsm-2-user-guide/hsm2-initial-provision-deploy-guide.html#pkcs11-rsa

**Comment by gilweis** [comment]
As you can see, the public key and cert are not public.

PKCS#11	YubiHSM 2	Rationale
CKA_PRIVATE

CK_TRUE

Login is always required

**Comment by alonbl** [comment]
I could not see how you placed them there... :)

**Comment by gilweis** [comment]
https://developers.yubico.com/YubiHSM2/Component_Reference/PKCS_11/

Section "PKCS#11 Attributes":
There are a number of attributes defined in PKCS#11 that do not translate to Capabilities of the YubiHSM 2 device, and are therefore treated as always having a fixed value.

**Comment by alonbl** [comment]
Bummer...
CKA_PRIVATE|CK_TRUE|Login is always required
You can modify the attribute of the certificate and add a flag that it is
private so that login will be done earlier.

**Comment by gilweis** [comment]
https://github.com/Yubico/yubihsm-shell/issues/336
It's look like pkcs11-tool ignoring the CKA_PRIVATE when writing certificates.
Why not give mycms-tool an option to force login for reading the objects (like --pin option in pkcs11-tool and like openssl pkcs11 engine and openssl pkcs11 provider)?

**Comment by alonbl** [comment]
This is what I referred to in my last comment.
Do you have anyone to workout a patch?

**Comment by gilweis** [comment]
I can test a patch (but not to workout a patch)

**Comment by gilweis** [comment]
Hi,
I tried to change the code (src/libmycms/mycms-certificate-driver-pkcs11.c) and do the login at __driver_load() function (if necessery) instead at __open_private_key() function.
It's working with this change :-)
It's ok to change the code in such way?

**Comment by alonbl** [comment]
Please test #11

**Comment by gilweis** [comment]
Thanks.
Tested (sign,verify,encrypt-add,encrypt) with:
SoftHSM and cert-is-private=0
YubiHSM and cert-is-private=1
YubiKey5 and cert-is-private=0
All the tests passed!

**Comment by alonbl** [comment]
Fixed with #11.
